### PR TITLE
Proxy: Add Parent Proxy feature

### LINKED ISSUE #2068: Parent Proxy feature
Asking for a ParentProxy feature.
At this time there is only one way to configure that in the opnsense filesystem. See: https://forum.opnsense.org/index.php?topic=6763.msg29327#msg29327

### PR DESCRIPTION
**Pull request by mimugmail**
Closes #2068 

- [x] 1. I'm unsure about the required fields in model, should they be better set to yes?
- [x] 2. The templating creates a good config (in my point of view), but it's not sent via Parent, perhaps you have an idea why?

Example:

```
root@OPNsense:~/core # cat /usr/local/etc/squid/pre-auth/parentproxy.conf
cache_peer 10.24.66.3 parent 8080 0 no-query default
acl ExcludePPDomains dstdomain test.de bla.de
acl ExcludePPIPs dst 1.1.1.1 2.2.2.2
cache_peer_access 10.24.66.3 deny ExcludePPDomains
cache_peer_access 10.24.66.3 deny ExcludePPIPs
cache_peer_access 10.24.66.3 allow all
never_direct allow ExcludePPDomains
never_direct allow ExcludePPIPs
never_direct deny all

```

Reference:
https://forum.opnsense.org/index.php?topic=6763.msg29212#msg29212

### CODE REVIEW

File: src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+cache_peer_access {{ OPNsense.proxy.general.parentproxy.host }} deny ExcludePPDomains
+{%   endif %}
+{%   if helpers.exists('OPNsense.proxy.general.parentproxy.localips') and OPNsense.proxy.general.parentproxy.localips != '' %}
+cache_peer_access {{ OPNsense.proxy.general.parentproxy.host }} deny ExcludePPIPs
+{%   endif %}
+cache_peer_access {{ OPNsense.proxy.general.parentproxy.host }} allow all
+{%   if helpers.exists('OPNsense.proxy.general.parentproxy.localdomains') and OPNsense.proxy.general.parentproxy.localdomains != '' %}
+never_direct allow ExcludePPDomains
+{%   endif %}
+{%   if helpers.exists('OPNsense.proxy.general.parentproxy.localips') and OPNsense.proxy.general.parentproxy.localips != '' %}
```
Comment (fabianfrz) [inline src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf]: you may group the settings of `OPNsense.proxy.general.parentproxy.localips` together in a single if statement

File: src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml [hunk 2]
```diff
@@ -182,6 +182,24 @@
...
+                <enabled type="BooleanField">
+                    <default>0</default>
+                    <Required>Y</Required>
+                </enabled>
+                <host type="HostnameField">
+                    <Required>N</Required>
+                </host>
+                <port type="PortField">
+                    <Required>N</Required>
+                </port>
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml]: you may want to use my SetIfContraint

Comment (mimugmail) [reply src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml]: Thanks, nice idea, I did it with DependConstraint

File: src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf (hunk 1 above)
Comment (mimugmail) [reply src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf]: Wouldn't this mixup the order if acl, cache_peer_access and never_direct when putting in a single if? 
Config would look a bit weird.

File: src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml [hunk 3]
```diff
@@ -182,6 +182,42 @@
...
+                                <field1>enabled</field1>
+                            </addFields>
+                        </check001>
+                    </Constraints>
+                </host>
+                <port type="PortField">
+                    <Required>N</Required>
+                    <Constraints>
+                        <check001>
+                            <ValidationMessage>A host must be set.</ValidationMessage>
```
Comment (fichtner) [inline src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml]: port

File: src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{% if helpers.exists('OPNsense.proxy.general.parentproxy.enabled') and OPNsense.proxy.general.parentproxy.enabled == '1' %}
```
Comment (fichtner) [inline src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf]: Normally what require Y does is it enforces this particular setting so we don’t have to tiptoe around helpers.exist all the time in the template. The same goes for constraints tied to this setting.

File: src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml (hunk 3 above)
Comment (mimugmail) [reply src/opnsense/mvc/app/models/OPNsense/Proxy/Proxy.xml]: fixed

File: src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf (hunk 4 above)
Comment (mimugmail) [reply src/opnsense/service/templates/OPNsense/Proxy/parentproxy.conf]: True, sorry

**Comment by mimugmail** [comment]
If you need a parent proxy for testing just ping me :)

**Comment by AdSchellevis** [comment]
@mimugmail I checked the logic in the ruleset, it looks good, but I haven't tested it on a parent server myself. It safe to add to our dev branch, if there are issues, they shouldn't be too hard to fix. Thanks!

**Pull request merged**